src/render/painter.js

```diff
@@ -804,6 +843,46 @@ class Painter {
         }
     }
 
+    prepareDrawProgram(context: Context, program: Program<*>, tileID: ?UnwrappedTileID) {
```

Comment:
Uniforms, buffers, attributes are all otherwise set in `program.draw(...)`. Would it be possible to make setting fog uniforms fit into that system?

Reply:
I investigated that initially and it seemed like the best option to not make our program class depend directly or indirectly on any fog details.

Fog could have its own dedicated system/class, but considering this was the only specific function that was doing some fog rendering logic it seemed appropriate as just a function.

Terrain has a very similar code path [here](https://github.com/mapbox/mapbox-gl-js/blob/main/src/terrain/terrain.js#L551_L600), where uniforms are set globally to the shader. Root properties that interact with a shader need to have some specific interaction/rendering logic and adding each of these details in our program class (or accomodating the interface for it) doesn't seem like the best option.

One more thing to consider is that not all `program.draw(...)` would need fog uniforms.